File: .agents/skills/rspack-binary-size/scripts/macro-expansion-stats.py
#!/usr/bin/env python3
import argparse
import hashlib
import json
import os
import queue
import re
import subprocess
import threading
import time
from pathlib import Path
from urllib.parse import quote


SKIP_DIRS = {".git", "target", "node_modules", ".turbo", ".next", "dist", "coverage"}

BUILTIN_ATTRS = {
    "allow",
    "cfg",
    "cfg_attr",
    "clippy",
    "cold",
    "deny",
    "deprecated",
    "doc",
    "expect",
    "forbid",
    "inline",
    "link",
    "must_use",
    "no_mangle",
    "non_exhaustive",
    "path",
    "repr",
    "rustfmt",
    "should_panic",
    "test",
    "warn",
}

ATTR_RE = re.compile(r"#\s*!\s*\[\s*([A-Za-z_][A-Za-z0-9_:]*)|#\s*\[\s*([A-Za-z_][A-Za-z0-9_:]*)")
CALL_RE = re.compile(r"\b([A-Za-z_][A-Za-z0-9_:]*)!\s*[\(\{\[]")
DERIVE_RE = re.compile(r"#\s*\[\s*derive\s*\((.*?)\)\s*\]", re.S)
IDENT_RE = re.compile(r"[A-Za-z_][A-Za-z0-9_:]*")


def uri_for(path: Path) -> str:
    return "file://" + quote(str(path.resolve()))


def normalize_macro_name(name: str) -> str:
    return name.split("::")[-1]


def iter_rust_files(root: Path, scan_roots: list[str], max_bytes: int):
    for scan_root in scan_roots:
        base = root / scan_root
        if not base.exists():
            continue
        for dirpath, dirnames, filenames in os.walk(base):
            dirnames[:] = [d for d in dirnames if d not in SKIP_DIRS]
            for filename in filenames:
                if not filename.endswith(".rs"):
                    continue
                path = Path(dirpath) / filename
                try:
                    if path.stat().st_size > max_bytes:
                        continue
                except OSError:
                    continue
                yield path


def line_col(text: str, offset: int) -> tuple[int, int]:
    line = text.count("\n", 0, offset)
    prev = text.rfind("\n", 0, offset)
    col = offset if prev < 0 else offset - prev - 1
    return line, col


def line_at(text: str, line: int) -> str:
    lines = text.splitlines()
    return lines[line] if 0 <= line < len(lines) else ""


def collect_candidates(root: Path, scan_roots: list[str], max_bytes: int):
    candidates = []
    per_macro: dict[tuple[str, str], dict[str, object]] = {}

    def add(kind: str, name: str, path: Path, line: int, col: int, snippet: str):
        norm = normalize_macro_name(name)
        key = (kind, norm)
        rel = str(path.relative_to(root))
        entry = per_macro.setdefault(key, {"count": 0, "files": set()})
        entry["count"] = int(entry["count"]) + 1
        entry["files"].add(rel)
        candidates.append(
            {
                "kind": kind,
                "name": norm,
                "raw_name": name,
                "path": path,
                "rel": rel,
                "line": line,
                "character": col,
                "snippet": snippet.strip()[:240],
            }
        )

    for path in iter_rust_files(root, scan_roots, max_bytes):
        try:
            text = path.read_text(encoding="utf-8")
        except UnicodeDecodeError:
            text = path.read_text(encoding="utf-8", errors="replace")
        except OSError:
            continue

        for match in DERIVE_RE.finditer(text):
            body = match.group(1)
            for item in IDENT_RE.finditer(body):
                name = item.group(0)
                line, col = line_col(text, match.start(1) + item.start())
                add("derive", name, path, line, col, match.group(0).replace("\n", " "))

        for match in ATTR_RE.finditer(text):
            name = match.group(1) or match.group(2)
            if not name:
                continue
            norm = normalize_macro_name(name)
            if norm in BUILTIN_ATTRS:
                continue
            line, col = line_col(text, match.start())
            add("attribute", name, path, line, col, line_at(text, line))

        for match in CALL_RE.finditer(text):
            name = match.group(1)
            line, col = line_col(text, match.start(1))
            add("function_like", name, path, line, col, line_at(text, line))

    return candidates, per_macro


class RustAnalyzerClient:
    def __init__(self, root: Path, analyzer: str):
        self.root = root
        self.proc = subprocess.Popen(
            [analyzer],
            cwd=str(root),
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=False,
        )
        self.next_id = 1
        self.response_queue: queue.Queue[object] = queue.Queue()
        self.notifications = []
        self.errors = []
        self.reader = threading.Thread(target=self._read_loop, daemon=True)
        self.reader.start()

    def _read_loop(self):
        try:
            while True:
                headers = {}
                while True:
                    line = self.proc.stdout.readline()
                    if not line:
                        return
                    if line in (b"\r\n", b"\n"):
                        break
                    key, value = line.decode("ascii").split(":", 1)
                    headers[key.lower()] = value.strip()
                length = int(headers.get("content-length", "0"))
                if length <= 0:
                    continue
                payload = self.proc.stdout.read(length)
                self.response_queue.put(json.loads(payload.decode("utf-8")))
        except Exception as error:
            self.errors.append(f"reader failed: {error}")

    def send(self, message: dict[str, object]):
        encoded = json.dumps(message, separators=(",", ":")).encode("utf-8")
        header = f"Content-Length: {len(encoded)}\r\n\r\n".encode("ascii")
        assert self.proc.stdin is not None
        self.proc.stdin.write(header + encoded)
        self.proc.stdin.flush()

    def request(self, method: str, params: object, timeout: float = 60.0):
        request_id = self.next_id
        self.next_id += 1
        self.send({"jsonrpc": "2.0", "id": request_id, "method": method, "params": params})
        deadline = time.monotonic() + timeout
        while time.monotonic() < deadline:
            try:
                message = self.response_queue.get(timeout=0.25)
            except queue.Empty:
                if self.proc.poll() is not None:
                    raise RuntimeError(f"rust-analyzer exited with code {self.proc.returncode}")
                continue
            if "id" in message and message["id"] == request_id:
                if "error" in message:
                    raise RuntimeError(json.dumps(message["error"], ensure_ascii=False))
                return message.get("result")
            self.notifications.append(message)
        raise TimeoutError(f"timed out waiting for {method}")

    def notify(self, method: str, params: object):
        self.send({"jsonrpc": "2.0", "method": method, "params": params})

    def initialize(self):
        result = self.request(
            "initialize",
            {
                "processId": os.getpid(),
                "rootUri": uri_for(self.root),
                "capabilities": {},
                "workspaceFolders": [{"uri": uri_for(self.root), "name": self.root.name}],
            },
            timeout=120.0,
        )
        self.notify("initialized", {})
        return result

    def shutdown(self):
        try:
            self.request("shutdown", None, timeout=10.0)
            self.notify("exit", None)
        except Exception as error:
            self.errors.append(f"shutdown failed: {error}")
        finally:
            try:
                self.proc.terminate()
            except Exception:
                pass


def write_tsv(path: Path, rows: list[list[object]], header: list[str]):
    with path.open("w", encoding="utf-8") as output:
        output.write("\t".join(header) + "\n")
        for row in rows:
            output.write("\t".join(str(value).replace("\t", " ") for value in row) + "\n")


def write_scan_reports(out: Path, candidates, per_macro):
    summary_rows = []
    for (kind, name), data in sorted(
        per_macro.items(),
        key=lambda item: (-int(item[1]["count"]), item[0][0], item[0][1]),
    ):
        summary_rows.append([kind, name, data["count"], len(data["files"])])
    write_tsv(out / "macro-occurrences.tsv", summary_rows, ["kind", "name", "count", "files"])

    location_rows = [
        [item["kind"], item["name"], item["rel"], item["line"] + 1, item["character"] + 1, item["snippet"]]
        for item in candidates
    ]
    write_tsv(out / "macro-locations.tsv", location_rows, ["kind", "name", "file", "line", "character", "snippet"])

    per_file: dict[str, int] = {}
    for item in candidates:
        per_file[item["rel"]] = per_file.get(item["rel"], 0) + 1
    file_rows = [[count, rel] for rel, count in sorted(per_file.items(), key=lambda item: (-item[1], item[0]))]
    write_tsv(out / "macro-files.tsv", file_rows, ["count", "file"])


def expand_with_rust_analyzer(root: Path, out: Path, candidates):
    analyzer = os.environ.get("RUST_ANALYZER", "rust-analyzer")
    limit = int(os.environ.get("RA_EXPAND_LIMIT", "200"))
    pattern = os.environ.get("RA_EXPAND_FILTER", "")
    name_filter = re.compile(pattern) if pattern else None
    timeout = float(os.environ.get("RA_EXPAND_TIMEOUT", "45"))

    selected = []
    for item in candidates:
        if name_filter and not name_filter.search(item["name"]):
            continue
        selected.append(item)
        if len(selected) >= limit:
            break

    expansions_dir = out / "rust-analyzer-expansions"
    expansions_dir.mkdir(exist_ok=True)
    rows = []
    errors = []
    opened = set()
    client = RustAnalyzerClient(root, analyzer)
    try:
        client.initialize()
        for item in selected:
            path = item["path"]
            uri = uri_for(path)
            if uri not in opened:
                client.notify(
                    "textDocument/didOpen",
                    {
                        "textDocument": {
                            "uri": uri,
                            "languageId": "rust",
                            "version": 1,
                            "text": path.read_text(encoding="utf-8", errors="replace"),
                        }
                    },
                )
                opened.add(uri)

            params = {
                "textDocument": {"uri": uri},
                "position": {"line": item["line"], "character": item["character"]},
            }
            try:
                result = client.request("rust-analyzer/expandMacro", params, timeout=timeout)
            except Exception as error:
                errors.append(f"{item['rel']}:{item['line'] + 1}:{item['character'] + 1} {item['name']}: {error}")
                rows.append(["error", item["kind"], item["name"], item["rel"], item["line"] + 1, 0, 0, ""])
                continue

            if not result:
                rows.append(["empty", item["kind"], item["name"], item["rel"], item["line"] + 1, 0, 0, ""])
                continue

            expansion = result.get("expansion", "") if isinstance(result, dict) else str(result)
            macro_name = result.get("name", item["name"]) if isinstance(result, dict) else item["name"]
            digest = hashlib.sha256(expansion.encode("utf-8")).hexdigest()[:16]
            filename = f"{len(rows):05d}-{item['name']}-{digest}.rs".replace("/", "_")
            (expansions_dir / filename).write_text(expansion, encoding="utf-8")
            rows.append(
                [
                    "ok",
                    item["kind"],
                    macro_name,
                    item["rel"],
                    item["line"] + 1,
                    len(expansion.encode("utf-8")),
                    expansion.count("\n") + 1 if expansion else 0,
                    filename,
                ]
            )
    finally:
        client.shutdown()

    write_tsv(
        out / "rust-analyzer-expansions.tsv",
        rows,
        ["status", "kind", "name", "file", "line", "bytes", "lines", "expansion_file"],
    )
    if errors or client.errors:
        (out / "rust-analyzer-errors.txt").write_text("\n".join(errors + client.errors) + "\n", encoding="utf-8")


def cargo_expand(root: Path, out: Path):
    crates = os.environ.get("CARGO_EXPAND_CRATES", "rspack_binding_api").split()
    markers = [
        "__napi",
        "FromNapiValue",
        "ToNapiValue",
        "ValidateNapiValue",
        "TypeName",
        "CallbackInfo",
        "register_class",
        "register_module_export",
        "ThreadsafeFunction",
        "Either",
    ]
    rows = []
    for crate in crates:
        expanded = out / f"cargo-expand-{crate}.rs"
        stderr = out / f"cargo-expand-{crate}.stderr.txt"
        with expanded.open("w", encoding="utf-8") as stdout, stderr.open("w", encoding="utf-8") as err:
            proc = subprocess.run(["cargo", "expand", "-p", crate], cwd=root, stdout=stdout, stderr=err, check=False)
        if proc.returncode != 0:
            rows.append([crate, "error", 0, 0, ""])
            continue
        text = expanded.read_text(encoding="utf-8", errors="replace")
        marker_counts = ",".join(f"{marker}={text.count(marker)}" for marker in markers)
        rows.append([crate, "ok", len(text.encode("utf-8")), text.count("\n") + 1, marker_counts])
    write_tsv(out / "cargo-expand-summary.tsv", rows, ["crate", "status", "bytes", "lines", "marker_counts"])


def main():
    parser = argparse.ArgumentParser(description="Collect generic Rust macro usage and optional macro expansion sizes.")
    parser.add_argument("--root", default=".", help="repository root")
    parser.add_argument("--out", required=True, help="output directory")
    args = parser.parse_args()

    root = Path(args.root).resolve()
    out = Path(args.out).resolve()
    out.mkdir(parents=True, exist_ok=True)

    scan_roots = os.environ.get("MACRO_SCAN_ROOTS", "crates packages .agents").split()
    max_bytes = int(os.environ.get("MACRO_SCAN_MAX_BYTES", str(1024 * 1024)))

    candidates, per_macro = collect_candidates(root, scan_roots, max_bytes)
    write_scan_reports(out, candidates, per_macro)

    backend = os.environ.get("MACRO_EXPAND_BACKEND", "none")
    if backend == "rust-analyzer":
        try:
            expand_with_rust_analyzer(root, out, candidates)
        except FileNotFoundError:
            (out / "rust-analyzer.skipped.txt").write_text(
                "rust-analyzer was not found. Set RUST_ANALYZER=/path/to/rust-analyzer or use MACRO_EXPAND_BACKEND=none.\n",
                encoding="utf-8",
            )
        except Exception as error:
            (out / "rust-analyzer.failed.txt").write_text(f"{error}\n", encoding="utf-8")
    elif backend == "cargo-expand":
        try:
            cargo_expand(root, out)
        except FileNotFoundError as error:
            (out / "cargo-expand.skipped.txt").write_text(f"{error}\n", encoding="utf-8")
    elif backend != "none":
        (out / "backend.skipped.txt").write_text(f"unknown MACRO_EXPAND_BACKEND={backend}\n", encoding="utf-8")

    print(f"macro candidates: {len(candidates)}")


if __name__ == "__main__":
    main()

File: .agents/skills/rspack-binary-size/scripts/macro-expansion-stats.mjs
#!/usr/bin/env node
import { $ } from 'zx';
import { createHash } from 'node:crypto';
import { spawn } from 'node:child_process';
import {
  existsSync,
  mkdirSync,
  readdirSync,
  readFileSync,
  statSync,
  writeFileSync,
} from 'node:fs';
import path from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';

$.quiet = true;

const skipDirs = new Set([
  '.git',
  'target',
  'node_modules',
  '.turbo',
  '.next',
  'dist',
  'coverage',
]);
const builtinAttrs = new Set([
  'allow',
  'cfg',
  'cfg_attr',
  'clippy',
  'cold',
  'deny',
  'deprecated',
  'doc',
  'expect',
  'forbid',
  'inline',
  'link',
  'must_use',
  'no_mangle',
  'non_exhaustive',
  'path',
  'repr',
  'rustfmt',
  'should_panic',
  'test',
  'warn',
]);

const attrRe =
  /#\s*!\s*\[\s*([A-Za-z_][A-Za-z0-9_:]*)|#\s*\[\s*([A-Za-z_][A-Za-z0-9_:]*)/g;
const callRe = /\b([A-Za-z_][A-Za-z0-9_:]*)!\s*[[({]/g;
const deriveRe = /#\s*\[\s*derive\s*\((.*?)\)\s*\]/gs;
const identRe = /[A-Za-z_][A-Za-z0-9_:]*/g;

async function repoRoot() {
  try {
    return (await $`git rev-parse --show-toplevel`).stdout.trim();
  } catch {
    return process.cwd();
  }
}

function positionalArgs() {
  const args = process.argv.slice(2);
  const scriptPath = fileURLToPath(import.meta.url);
  if (args[0] && path.resolve(args[0]) === scriptPath) args.shift();
  if (args[0] === '--') args.shift();
  return args;
}

function stamp() {
  const d = new Date();
  const pad = (n) => String(n).padStart(2, '0');
  return `${d.getFullYear()}${pad(d.getMonth() + 1)}${pad(d.getDate())}-${pad(d.getHours())}${pad(d.getMinutes())}${pad(d.getSeconds())}`;
}

function normalizeMacroName(name) {
  return name.split('::').at(-1);
}

function lineCol(text, offset) {
  let line = 0;
  let last = -1;
  for (let i = 0; i < offset; i++) {
    if (text.charCodeAt(i) === 10) {
      line++;
      last = i;
    }
  }
  return [line, offset - last - 1];
}

function lineAt(text, line) {
  return text.split(/\r?\n/)[line] ?? '';
}

function* walkRustFiles(root, scanRoots, maxBytes) {
  function* walk(dir) {
    let entries;
    try {
      entries = readdirSync(dir, { withFileTypes: true });
    } catch {
      return;
    }
    for (const entry of entries) {
      if (entry.isDirectory()) {
        if (!skipDirs.has(entry.name)) yield* walk(path.join(dir, entry.name));
      } else if (entry.isFile() && entry.name.endsWith('.rs')) {
        const file = path.join(dir, entry.name);
        try {
          if (statSync(file).size <= maxBytes) yield file;
        } catch {}
      }
    }
  }

  for (const scanRoot of scanRoots) {
    const base = path.join(root, scanRoot);
    if (existsSync(base)) yield* walk(base);
  }
}

function collectCandidates(root, scanRoots, maxBytes) {
  const candidates = [];
  const perMacro = new Map();

  function add(kind, name, file, line, character, snippet) {
    const normalized = normalizeMacroName(name);
    const rel = path.relative(root, file);
    const key = `${kind}\t${normalized}`;
    const current = perMacro.get(key) ?? {
      kind,
      name: normalized,
      count: 0,
      files: new Set(),
    };
    current.count++;
    current.files.add(rel);
    perMacro.set(key, current);
    candidates.push({
      kind,
      name: normalized,
      rawName: name,
      file,
      rel,
      line,
      character,
      snippet: snippet.trim().slice(0, 240),
    });
  }

  for (const file of walkRustFiles(root, scanRoots, maxBytes)) {
    const text = readFileSync(file, 'utf8');

    for (const match of text.matchAll(deriveRe)) {
      const body = match[1] ?? '';
      for (const item of body.matchAll(identRe)) {
        const [line, character] = lineCol(
          text,
          match.index + match[0].indexOf(body) + item.index,
        );
        add(
          'derive',
          item[0],
          file,
          line,
          character,
          match[0].replace(/\s+/g, ' '),
        );
      }
    }

    for (const match of text.matchAll(attrRe)) {
      const name = match[1] || match[2];
      if (!name) continue;
      const normalized = normalizeMacroName(name);
      if (builtinAttrs.has(normalized)) continue;
      const [line, character] = lineCol(text, match.index);
      add('attribute', name, file, line, character, lineAt(text, line));
    }

    for (const match of text.matchAll(callRe)) {
      const [line, character] = lineCol(text, match.index);
      add('function_like', match[1], file, line, character, lineAt(text, line));
    }
  }

  return { candidates, perMacro };
}

function writeTsv(file, header, rows) {
  const escape = (value) =>
    String(value).replaceAll('\t', ' ').replaceAll('\n', ' ');
  writeFileSync(
    file,
    `${header.join('\t')}\n${rows.map((row) => row.map(escape).join('\t')).join('\n')}\n`,
  );
}

function writeScanReports(outDir, candidates, perMacro) {
  const macroRows = [...perMacro.values()]
    .sort(
      (a, b) =>
        b.count - a.count ||
        a.kind.localeCompare(b.kind) ||
        a.name.localeCompare(b.name),
    )
    .map((item) => [item.kind, item.name, item.count, item.files.size]);
  writeTsv(
    path.join(outDir, 'macro-occurrences.tsv'),
    ['kind', 'name', 'count', 'files'],
    macroRows,
  );

  writeTsv(
    path.join(outDir, 'macro-locations.tsv'),
    ['kind', 'name', 'file', 'line', 'character', 'snippet'],
    candidates.map((item) => [
      item.kind,
      item.name,
      item.rel,
      item.line + 1,
      item.character + 1,
      item.snippet,
    ]),
  );

  const perFile = new Map();
  for (const item of candidates)
    perFile.set(item.rel, (perFile.get(item.rel) ?? 0) + 1);
  const fileRows = [...perFile.entries()]
    .sort((a, b) => b[1] - a[1] || a[0].localeCompare(b[0]))
    .map(([file, count]) => [count, file]);
  writeTsv(path.join(outDir, 'macro-files.tsv'), ['count', 'file'], fileRows);
}

class RustAnalyzerClient {
  constructor(root, analyzer) {
    this.root = root;
    this.nextId = 1;
    this.pending = new Map();
    this.notifications = [];
    this.errors = [];
    this.buffer = Buffer.alloc(0);
    this.proc = spawn(analyzer, [], {
      cwd: root,
      stdio: ['pipe', 'pipe', 'pipe'],
    });
    this.proc.stdout.on('data', (chunk) => this.onData(chunk));
    this.proc.stderr.on('data', (chunk) => this.errors.push(chunk.toString()));
  }

  onData(chunk) {
    this.buffer = Buffer.concat([this.buffer, chunk]);
    while (true) {
      const headerEnd = this.buffer.indexOf('\r\n\r\n');
      if (headerEnd < 0) return;
      const header = this.buffer.slice(0, headerEnd).toString('ascii');
      const lengthMatch = /content-length:\s*(\d+)/i.exec(header);
      if (!lengthMatch) throw new Error(`missing Content-Length in ${header}`);
      const length = Number(lengthMatch[1]);
      const bodyStart = headerEnd + 4;
      if (this.buffer.length < bodyStart + length) return;
      const body = this.buffer
        .slice(bodyStart, bodyStart + length)
        .toString('utf8');
      this.buffer = this.buffer.slice(bodyStart + length);
      const message = JSON.parse(body);
      if (message.id && this.pending.has(message.id)) {
        const { resolve, reject } = this.pending.get(message.id);
        this.pending.delete(message.id);
        if (message.error) reject(new Error(JSON.stringify(message.error)));
        else resolve(message.result);
      } else {
        this.notifications.push(message);
      }
    }
  }

  send(message) {
    const body = Buffer.from(JSON.stringify(message), 'utf8');
    this.proc.stdin.write(`Content-Length: ${body.length}\r\n\r\n`);
    this.proc.stdin.write(body);
  }

  request(method, params, timeoutMs = 60_000) {
    const id = this.nextId++;
    this.send({ jsonrpc: '2.0', id, method, params });
    return new Promise((resolve, reject) => {
      const timeout = setTimeout(() => {
        this.pending.delete(id);
        reject(new Error(`timed out waiting for ${method}`));
      }, timeoutMs);
      this.pending.set(id, {
        resolve: (value) => {
          clearTimeout(timeout);
          resolve(value);
        },
        reject: (error) => {
          clearTimeout(timeout);
          reject(error);
        },
      });
    });
  }

  notify(method, params) {
    this.send({ jsonrpc: '2.0', method, params });
  }

  async initialize() {
    const result = await this.request(
      'initialize',
      {
        processId: process.pid,
        rootUri: pathToFileURL(this.root).href,
        capabilities: {},
        workspaceFolders: [
          {
            uri: pathToFileURL(this.root).href,
            name: path.basename(this.root),
          },
        ],
      },
      120_000,
    );
    this.notify('initialized', {});
    return result;
  }

  async shutdown() {
    try {
      await this.request('shutdown', null, 10_000);
      this.notify('exit', null);
    } catch (error) {
      this.errors.push(`shutdown failed: ${error.message}`);
    } finally {
      this.proc.kill();
    }
  }
}

async function expandWithRustAnalyzer(root, outDir, candidates) {
  const analyzer = process.env.RUST_ANALYZER || 'rust-analyzer';
  const limit = Number(process.env.RA_EXPAND_LIMIT || 200);
  const filter = process.env.RA_EXPAND_FILTER
    ? new RegExp(process.env.RA_EXPAND_FILTER)
    : null;
  const timeoutMs = Number(process.env.RA_EXPAND_TIMEOUT || 45) * 1000;
  const selected = [];
  for (const item of candidates) {
    if (filter && !filter.test(item.name)) continue;
    selected.push(item);
    if (selected.length >= limit) break;
  }

  const expansionsDir = path.join(outDir, 'rust-analyzer-expansions');
  mkdirSync(expansionsDir, { recursive: true });
  const rows = [];
  const errors = [];
  const opened = new Set();
  const client = new RustAnalyzerClient(root, analyzer);
  try {
    await client.initialize();
    for (const item of selected) {
      const uri = pathToFileURL(item.file).href;
      if (!opened.has(uri)) {
        client.notify('textDocument/didOpen', {
          textDocument: {
            uri,
            languageId: 'rust',
            version: 1,
            text: readFileSync(item.file, 'utf8'),
          },
        });
        opened.add(uri);
      }
      try {
        const result = await client.request(
          'rust-analyzer/expandMacro',
          {
            textDocument: { uri },
            position: { line: item.line, character: item.character },
          },
          timeoutMs,
        );
        if (!result) {
          rows.push([
            'empty',
            item.kind,
            item.name,
            item.rel,
            item.line + 1,
            0,
            0,
            '',
          ]);
          continue;
        }
        const expansion =
          typeof result === 'object' ? result.expansion || '' : String(result);
        const macroName =
          typeof result === 'object' ? result.name || item.name : item.name;
        const digest = createHash('sha256')
          .update(expansion)
          .digest('hex')
          .slice(0, 16);
        const filename =
          `${String(rows.length).padStart(5, '0')}-${item.name}-${digest}.rs`.replaceAll(
            '/',
            '_',
          );
        writeFileSync(path.join(expansionsDir, filename), expansion);
        rows.push([
          'ok',
          item.kind,
          macroName,
          item.rel,
          item.line + 1,
          Buffer.byteLength(expansion),
          expansion ? expansion.split('\n').length : 0,
          filename,
        ]);
      } catch (error) {
        errors.push(
          `${item.rel}:${item.line + 1}:${item.character + 1} ${item.name}: ${error.message}`,
        );
        rows.push([
          'error',
          item.kind,
          item.name,
          item.rel,
          item.line + 1,
          0,
          0,
          '',
        ]);
      }
    }
  } finally {
    await client.shutdown();
  }
  writeTsv(
    path.join(outDir, 'rust-analyzer-expansions.tsv'),
    [
      'status',
      'kind',
      'name',
      'file',
      'line',
      'bytes',
      'lines',
      'expansion_file',
    ],
    rows,
  );
  if (errors.length || client.errors.length)
    writeFileSync(
      path.join(outDir, 'rust-analyzer-errors.txt'),
      `${[...errors, ...client.errors].join('\n')}\n`,
    );
}

async function runCapture(command, args) {
  const result = await $({ nothrow: true })`${command} ${args}`;
  return {
    code: result.exitCode,
    stdout: result.stdout,
    stderr: result.stderr,
  };
}

async function cargoExpand(root, outDir) {
  const crates = (process.env.CARGO_EXPAND_CRATES || 'rspack_binding_api')
    .split(/\s+/)
    .filter(Boolean);
  const markers = [
    '__napi',
    'FromNapiValue',
    'ToNapiValue',
    'ValidateNapiValue',
    'TypeName',
    'CallbackInfo',
    'register_class',
    'register_module_export',
    'ThreadsafeFunction',
    'Either',
  ];
  const rows = [];
  for (const crate of crates) {
    const result = await runCapture('cargo', ['expand', '-p', crate]);
    writeFileSync(
      path.join(outDir, `cargo-expand-${crate}.stderr.txt`),
      result.stderr || '',
    );
    if (result.code !== 0) {
      rows.push([crate, 'error', 0, 0, '']);
      continue;
    }
    const expanded = result.stdout;
    writeFileSync(path.join(outDir, `cargo-expand-${crate}.rs`), expanded);
    const markerCounts = markers
      .map((marker) => `${marker}=${expanded.split(marker).length - 1}`)
      .join(',');
    rows.push([
      crate,
      'ok',
      Buffer.byteLength(expanded),
      expanded.split('\n').length,
      markerCounts,
    ]);
  }
  writeTsv(
    path.join(outDir, 'cargo-expand-summary.tsv'),
    ['crate', 'status', 'bytes', 'lines', 'marker_counts'],
    rows,
  );
}

const root = await repoRoot();
const [outDirArg] = positionalArgs();
const outDir = path.resolve(
  outDirArg ||
    path.join(
      root,
      'target',
      'binary-size-reports',
      `${stamp()}-macro-expansion`,
    ),
);
mkdirSync(outDir, { recursive: true });
process.chdir(root);

const scanRoots = (process.env.MACRO_SCAN_ROOTS || 'crates packages .agents')
  .split(/\s+/)
  .filter(Boolean);
const maxBytes = Number(process.env.MACRO_SCAN_MAX_BYTES || 1024 * 1024);
const { candidates, perMacro } = collectCandidates(root, scanRoots, maxBytes);
writeScanReports(outDir, candidates, perMacro);

const backend = process.env.MACRO_EXPAND_BACKEND || 'none';
if (backend === 'rust-analyzer') {
  try {
    await expandWithRustAnalyzer(root, outDir, candidates);
  } catch (error) {
    writeFileSync(
      path.join(outDir, 'rust-analyzer.failed.txt'),
      `${error.message}\n`,
    );
  }
} else if (backend === 'cargo-expand') {
  try {
    await cargoExpand(root, outDir);
  } catch (error) {
    writeFileSync(
      path.join(outDir, 'cargo-expand.failed.txt'),
      `${error.message}\n`,
    );
  }
} else if (backend !== 'none') {
  writeFileSync(
    path.join(outDir, 'backend.skipped.txt'),
    `unknown MACRO_EXPAND_BACKEND=${backend}\n`,
  );
}

console.log(`macro candidates: ${candidates.length}`);
console.log(`macro expansion report: ${outDir}`);
